### PR TITLE
feat: Schedule: Add Date above sessions of the same date

### LINKED ISSUE #5773: Schedule: Add Date above sessions of the same date
As you can see in the below screenshot the website generator has a heading above sessions of one date. Please implement the same in the eventyay version.

Add dates above sessions in the "by time" view. Do not show it when sessions are sorted by title.

Screenshot from 2020-11-25 20-53-14

### PR DESCRIPTION
**Pull request by daretobedifferent18**
Fixes #5773

Responsive | ScreenShot
 --------------- | ---------------
Mobile(sort-time) | Screenshot from 2020-12-11 03-59-00
Desktop (sort-time) | Screen Shot 2020-12-11 at 03 49 49
Desktop (sort-title) | Screenshot from 2020-12-11 03-57-19

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: app/routes/public/sessions.js [hunk 1]
```diff
@@ -126,21 +127,38 @@ export default class SessionsRoute extends Route {
...
+      include      : 'track,speakers,session-type,microlocation',
+      filter       : filterOptions,
+      sort         : params.sort || 'starts-at',
+      perPage      : 6,
+      startingPage : 1,
+      perPageParam : 'page[size]',
+      pageParam    : 'page[number]'
+    });
+
+    let groupByDateSessions = groupBy(session.toArray(), s => s.startsAt);
```
Comment (iamareebjamal) [inline app/routes/public/sessions.js:144]: No need of doing it here, create a computed property in controller

**Comment by iamareebjamal** [comment]
These are a whole lotta sessions https://open-event-frontend-git-show-date-5773.eventyay.now.sh/e/ae73f280/schedule

**Comment by daretobedifferent18** [comment]
> These are a whole lotta sessions https://open-event-frontend-git-show-date-5773.eventyay.now.sh/e/ae73f280/schedule

Ok I will test on this.

**Comment by daretobedifferent18** [comment]
@iamareebjamal everything is working fine. Just need to fix infinityLoader. It is rotating infinitly

**Comment by maze-runnar** [comment]
sponsor like grouping is not possible here due to infinityModel.  you are trying to load all sessions and then grouping them, if I am getting it correct? So the session that are loading in first go will be grouped and then loader will keep loading or not load further.

**Comment by daretobedifferent18** [comment]
so @maze-runnar either we have to remove the infinity loader

**Comment by daretobedifferent18** [comment]
> no, removing infinity loader is never an option.

so could you suggest me what could be the alternative way

**Comment by iamareebjamal** [comment]
You should only group loaded sessions, not triggering the infinity loader. Check how

**Comment by daretobedifferent18** [comment]
> You should only group loaded sessions, not triggering the infinity loader. Check how

@iamareebjamal could you pls guide me where to look at.
I have tried but not able to get for it

**Comment by daretobedifferent18** [comment]
> You should only group loaded sessions, not triggering the infinity loader. Check how

@iamareebjamal what i found is that group sessions are not grouped after infinter loader or rest session are loaded, I think we should observer function which observe on session

**Comment by daretobedifferent18** [comment]
I want to know that can we write observer funtion in controller file?

**Comment by iamareebjamal** [comment]
Will check ASAP, please take other issues for now

**Comment by daretobedifferent18** [comment]
@iamareebjamal Pls have a look. I have resolved the bug that was coming

**Comment by daretobedifferent18** [comment]
@iamareebjamal Pls have a look

**Comment by mariobehling** [comment]
Seems to be working nicely where I am testing currently e.g. here https://open-event-frontend-git-show-date-5773.eventyay.now.sh/e/3689bc98/schedule?track=Open%20Data%2C%20Internet%2C%20Community

**Comment by iamareebjamal** [comment]
Thanks, great work. Functionality is working, I'll suggest some code changes soon

**Comment by mariobehling** [comment]
So, unfortunately the times of the FOSSASIA Summit 2019 got messed up at some point. The sessions start at 5pm in the afternoon, but the correct time would be 9AM in the morning.

In regards to this issue I see that sessions of 11:40 PM are at the same date as 12:50 AM. Expected is that sessions that start after 23:59 PM show up under the new date.

Screenshot from 2020-12-18 19-55-39

**Comment by mariobehling** [comment]
Please compare here http://2019.fossasia.org/event/schedule.html

File: app/controllers/public/sessions.js [hunk 2]
```diff
@@ -13,6 +14,32 @@ export default class SessionsController extends Controller {
   timezone = null;
   preserveScrollPosition = true;
 
+  @computed('model.session.@each')
+  get groupByDateSessions() {
+    let sessions = groupBy(this.model.session.toArray(), s => moment(s.startsAt).format('dddd, Do MMMM'));
```
Comment (iamareebjamal) [inline app/controllers/public/sessions.js:19]: To fix what @mariobehling mentioned, please use `this.timezone` to convert the date to proper timezone before grouping

Comment (daretobedifferent18) [reply app/controllers/public/sessions.js:19]: adding like ```moment(s.startsAt).tz(this.timezone).format('dddd, Do MMMM'))``` give error  
TypeError: _moment.default(...).tz(...) is undefined

Comment (iamareebjamal) [reply app/controllers/public/sessions.js:19]: moment.tz(date, tz)

**Comment by daretobedifferent18** [comment]
@iamareebjamal Done

File: app/controllers/public/sessions.js [hunk 3]
```diff
@@ -13,6 +14,32 @@ export default class SessionsController extends Controller {
   timezone = null;
   preserveScrollPosition = true;
 
+  @computed('model.session.@each', 'timezone')
+  get groupByDateSessions() {
+    let sessions = groupBy(this.model.session.toArray(), s => moment.tz(s.startsAt, this.timezone).format('dddd, Do MMMM'));
```
Comment (iamareebjamal) [inline app/controllers/public/sessions.js:19]: Do this in else below

File: app/controllers/public/sessions.js [hunk 4]
```diff
@@ -13,6 +14,32 @@ export default class SessionsController extends Controller {
...
+        date = null;
+      }
+
+      arr.push({
+        date,
+        'sessions': sessions[key]
+      });
+    }
+
+    return arr;
```
Comment (iamareebjamal) [inline app/controllers/public/sessions.js:40]:
```suggestion
    return Object.entries(sessions).map((date, sessions) => ({ date, sessions }));
```

And remove the loop from above

Comment (daretobedifferent18) [reply app/controllers/public/sessions.js]: This was not working properly so i modified little bit.
Pls see the changes made

File: app/controllers/public/sessions.js [hunk 5]
```diff
@@ -16,28 +16,20 @@ export default class SessionsController extends Controller {
...
     }
 
-    const arr = [];
-
-    for (const key in sessions) {
-      let date = key;
-
-      if (date === 'undefined') {
-        date = null;
+    return Object.entries(sessions).map(groupedSessions => (
```
Comment (iamareebjamal) [inline app/controllers/public/sessions.js:27]:
```suggestion
    return Object.entries(sessions).map(([date, sessions]) => (
```

File: app/controllers/public/sessions.js [hunk 6]
```diff
@@ -16,28 +16,20 @@ export default class SessionsController extends Controller {
...
+      {
+        'date'     : groupedSessions[0] === 'undefined' ? null : groupedSessions[0],
```
Comment (iamareebjamal) [inline app/controllers/public/sessions.js:29]:
```suggestion
        date: date === 'undefined' ? null : date,
```

File: app/controllers/public/sessions.js [hunk 7]
```diff
@@ -16,28 +16,20 @@ export default class SessionsController extends Controller {
...
+        'sessions' : groupedSessions[1]
```
Comment (iamareebjamal) [inline app/controllers/public/sessions.js:30]:
```suggestion
        sessions
```

File: app/templates/public/sessions.hbs [hunk 8]
```diff
@@ -53,15 +53,19 @@
...
+  {{#each this.groupByDateSessions as |group|}}
+    <h3 class="ui header" style="margin-bottom: -20px; color: rgba(0, 0, 0, 0.7);">{{t-var group.date}}</h3>
     <table class="ui very basic table">
-      <thead class="full-width">
-        <tr>
-          <th class="two wide" style="border: none;"></th>
-          <th style="border: none;"></th>
-        </tr>
-      </thead>
+      {{#unless this.device.isMobile}}
```
Comment (iamareebjamal) [inline app/templates/public/sessions.hbs:59]: Why?

File: app/templates/public/sessions.hbs [hunk 9]
```diff
@@ -53,15 +53,19 @@
 
 <div class="mt-8">
   {{#if this.model.session}}
+  {{#each this.groupByDateSessions as |group|}}
+    <h3 class="ui header" style="margin-bottom: -20px; color: rgba(0, 0, 0, 0.7);">{{t-var group.date}}</h3>
```
Comment (iamareebjamal) [inline app/templates/public/sessions.hbs:57]: Use classes instead of styles

Comment (iamareebjamal) [inline app/templates/public/sessions.hbs:57]: Date is not translatable, remove t-var

File: app/templates/public/sessions.hbs (hunk 8 above)
Comment (daretobedifferent18) [reply app/templates/public/sessions.hbs:59]: Without change | withchange
---- | --------
Screenshot from 2020-12-19 05-54-04 | Screenshot from 2020-12-19 05-55-26


due to styling for mobile view which is not good without adding this line of code

Comment (iamareebjamal) [reply app/templates/public/sessions.hbs:59]: With change, there is no padding on Monday, 4th January

Comment (daretobedifferent18) [reply app/templates/public/sessions.hbs:59]: U you say i can revert this change, but line below Monday does not looks gud

Comment (iamareebjamal) [reply app/templates/public/sessions.hbs:59]: There are other ways to remove the gap without removing all padding and margin

**Comment by iamareebjamal** [comment]
image

Heading is lighter than subheading, it should be opposite

**Comment by daretobedifferent18** [comment]
@iamareebjamal Pls see the changes made in styling and let me know to make any changes more

**Comment by mariobehling** [comment]
> image
> 
> Heading is lighter than subheading (time), it should be opposite

@daretobedifferent18 @iamareebjamal The gray font we use is often difficult to read. Can we just keep the dark color for the time and make the date in the same font color? I dont think we need too many different font colors. What do you think?

File: app/templates/public/sessions.hbs [hunk 10]
```diff
@@ -53,6 +53,8 @@
 
 <div class="mt-8">
   {{#if this.model.session}}
+  {{#each this.groupByDateSessions as |group|}}
+    <h3 class="ui header {{if this.device.isMobile 'mb--8' 'mb--4'}}">{{t group.date}}</h3>
```
Comment (iamareebjamal) [inline app/templates/public/sessions.hbs:57]: t doesn't work with dynamic data, and secondly, dates are not translatable, as mentioned previously

File: app/templates/public/sessions.hbs [hunk 11]
```diff
@@ -61,10 +63,10 @@
         </tr>
       </thead>
       <tbody>
-        {{#each this.model.session as |session|}}
+        {{#each group.sessions as |session|}}
           <tr>
             <td class="top aligned {{unless this.device.isMobile 'centered text'}}" style="border-top: none;">
-              <h4 class="ui header {{unless this.device.isMobile 'pt-4'}}" style="color: rgba(0, 0, 0, 0.7);">
+              <h4 class="ui header {{unless this.device.isMobile 'pt-4'}} text muted">
```
Comment (iamareebjamal) [inline app/templates/public/sessions.hbs:69]: Keep it black

File: app/controllers/public/sessions.js [hunk 12]
```diff
@@ -13,6 +14,24 @@ export default class SessionsController extends Controller {
...
+    } else {
+      sessions = groupBy(this.model.session.toArray(), s => moment.tz(s.startsAt, this.timezone).format('dddd, Do MMMM'));
+    }
+
+    return Object.entries(sessions).map(([date, sessions]) => (
+      {
+        date: date === 'undefined' ? null : date,
+        sessions
+      }
+    ));
```
Comment (iamareebjamal) [inline app/controllers/public/sessions.js:32]:
```suggestion
    return Object.entries(sessions).map(([date, sessions]) => ({ date: date === 'undefined' ? null : date, sessions }));
```

**Review by iamareebjamal — APPROVED** [review]

**Pull request merged**